zeroconf/__init__.py

```diff
@@ -2766,69 +2767,77 @@ def handle_query(self, msg: DNSIncoming, addr: Optional[str], port: int) -> None
+                            out.add_answer(
+                                msg,
+                                DNSPointer(
+                                    service.type, _TYPE_PTR, _CLASS_IN, service.other_ttl, service.name
+                                ),
                             )
-                        )
-                        for address in service.addresses_by_version(IPVersion.All):
-                            type_ = _TYPE_AAAA if _is_v6_address(address) else _TYPE_A
+
+                            # Add recommended additional answers according to
+                            # https://tools.ietf.org/html/rfc6763#section-12.1.
                             out.add_additional_answer(
-                                DNSAddress(
-                                    service.server,
-                                    type_,
+                                DNSService(
+                                    service.name,
+                                    _TYPE_SRV,
                                     _CLASS_IN | _CLASS_UNIQUE,
                                     service.host_ttl,
-                                    address,
+                                    service.priority,
+                                    service.weight,
+                                    cast(int, service.port),
+                                    service.server,
                                 )
                             )
-            else:
-                try:
+                            out.add_additional_answer(
+                                DNSText(
+                                    service.name,
+                                    _TYPE_TXT,
+                                    _CLASS_IN | _CLASS_UNIQUE,
+                                    service.other_ttl,
+                                    service.text,
+                                )
+                            )
+                            for address in service.addresses_by_version(IPVersion.All):
+                                type_ = _TYPE_AAAA if _is_v6_address(address) else _TYPE_A
+                                out.add_additional_answer(
+                                    DNSAddress(
+                                        service.server,
+                                        type_,
+                                        _CLASS_IN | _CLASS_UNIQUE,
+                                        service.host_ttl,
+                                        address,
+                                    )
+                                )
+                else:
                     if out is None:
                         out = DNSOutgoing(_FLAGS_QR_RESPONSE | _FLAGS_AA)
 
+                    name_to_find = question.name.lower()
+
                     # Answer A record queries for any service addresses we know
                     if question.type in (_TYPE_A, _TYPE_ANY):
                         for service in self.services.values():
```

Comment:
Here I could add `self.serviceservers` under the service lock to make it a `dict` of types with the values as a list of `ServiceInfo` so we can do 

`self.serviceservers.get(name_to_find)` and then copy the list so we only copy on a hit.

Reply:
I'd need to pop it back out of both of them in `update_service` and put the new one in.